### ISSUE TITLE
Unable convert Falcon PEFT model : RuntimeError: shape '[71, 3, 64, 4544]' is invalid for input of size 21229568

### ISSUE THREAD

**Issue by amir1m** [intel/intel-extension-for-transformers#731]
Hello,
I have PEFT fine tuned Falcon model. Followed the steps as per the `[README](https://github.com/intel/intel-extension-for-transformers/tree/main/intel_extension_for_transformers/llm/runtime/graph)`

1. Merged PEFT adapter into the model : No errors
2. However, while converting using: `python scripts/convert.py --outtype f32 --outfile model_f32.bin model_peft` getting following error:
```
transformer.word_embeddings.weight 2 (65024, 4544)
Traceback (most recent call last):
  File "intel-extension-for-transformers/intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py", line 169, in <module>
    main()
  File "intel-extension-for-transformers/intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py", line 132, in main
    qkv = list_vars[src].view(
**RuntimeError: shape '[71, 3, 64, 4544]' is invalid for input of size 21229568**
```
3. If I comment out the IF block on line#132 in convert_falcon.py (just to test) and run the inference, i get following error:
```
n_head_kv should be 1 (7B) or 8 (40B) in Falcon model, rather 71 
error loading model: model.cpp: tensor 'transformer.h.0.self_attention.query_key_value.weight' has wrong shape; expected  4544 x 13632, got  4544 x  4672
model_init_from_file: failed to load model
```

Can someone help how to convert PEFT tuned Falcon model?

Thanks.

**Comment by zhentaoyu** [comment]
Hi, @amir1m, I tried to convert the `falcon-7b` model from [HF](https://huggingface.co/tiiuae/falcon-7b) and I didn't meet this problem. Here is my test env:
```bash
transformers                     4.33.1
torch                            2.0.1+cpu
itrex commit   adb109bad9d4fd83586989991543f5cc2b655b9e
```
command: `python intel-extension-for-transformers/intel_extension_for_transformers/llm/runtime/graph/scripts/convert.py --outtype f32 --outfile model-fp32.bin models--tiiuae--falcon-7b/`

and the output logs like:
```bash
WARNING: You are currently loading Falcon using legacy code contained in the model repository. Falcon has now been fully ported into the Hugging Face transformers library. For the most up-to-date and high-performance version of the Falcon model code, please update to the latest version of
 transformers and then load the model without the trust_remote_code=True argument.

cmd: ['python', PosixPath('/home/sdp/yzt/intel-extension-for-transformers/intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py'), '--outfile', PosixPath('model-fp32.bin'), '--outtype', 'f32', 'models--tiiuae--falcon-7b/snapshots/898df1396f35e447d5fe44e0a3ccaaaa69f3
0d36']

WARNING: You are currently loading Falcon using legacy code contained in the model repository. Falcon has now been fully ported into the Hugging Face transformers library. For the most up-to-date and high-performance version of the Falcon model code, please update to the latest version of
 transformers and then load the model without the trust_remote_code=True argument.

Loading model:  models--tiiuae--falcon-7b/snapshots/898df1396f35e447d5fe44e0a3ccaaaa69f30d36
Loading checkpoint shards: 100%|██████████████████████████████████████████████████████████████████████████████████| 2/2 [00:09<00:00,  4.96s/it]
Model loaded:  models--tiiuae--falcon-7b/snapshots/898df1396f35e447d5fe44e0a3ccaaaa69f30d36
transformer.word_embeddings.weight 2 (65024, 4544)
transformer.h.0.self_attention.query_key_value.weight 2 (4672, 4544)
transformer.h.0.self_attention.dense.weight 2 (4544, 4544)
transformer.h.0.mlp.dense_h_to_4h.weight 2 (18176, 4544)
transformer.h.0.mlp.dense_4h_to_h.weight 2 (4544, 18176)
transformer.h.0.input_layernorm.weight 1 (4544,)
transformer.h.0.input_layernorm.bias 1 (4544,)
transformer.h.1.self_attention.query_key_value.weight 2 (4672, 4544)
transformer.h.1.self_attention.dense.weight 2 (4544, 4544)
transformer.h.1.mlp.dense_h_to_4h.weight 2 (18176, 4544)

.......

transformer.h.30.self_attention.query_key_value.weight 2 (4672, 4544)
transformer.h.30.self_attention.dense.weight 2 (4544, 4544)
transformer.h.30.mlp.dense_h_to_4h.weight 2 (18176, 4544)
transformer.h.30.mlp.dense_4h_to_h.weight 2 (4544, 18176)
transformer.h.30.input_layernorm.weight 1 (4544,)
transformer.h.30.input_layernorm.bias 1 (4544,)
transformer.h.31.self_attention.query_key_value.weight 2 (4672, 4544)
transformer.h.31.self_attention.dense.weight 2 (4544, 4544)
transformer.h.31.mlp.dense_h_to_4h.weight 2 (18176, 4544)
transformer.h.31.mlp.dense_4h_to_h.weight 2 (4544, 18176)
transformer.h.31.input_layernorm.weight 1 (4544,)
transformer.h.31.input_layernorm.bias 1 (4544,)
transformer.ln_f.weight 1 (4544,)
transformer.ln_f.bias 1 (4544,)
lm_head.weight 2 (65024, 4544)
Done. Output file: model-fp32.bin
```

So, there is any difference between your fine-tuned model and the HF model except for weight values? And also, please supply your env info, thanks.

**Comment by amir1m** [comment]
Hi @zhentaoyu ,
I am running on the cloud/docker Jupyter environment :

1. CPU Details
```
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                48
On-line CPU(s) list:   0-47
Thread(s) per core:    2
Core(s) per socket:    24
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) Platinum 8167M CPU @ 2.00GHz
Stepping:              4
CPU MHz:               1995.339
BogoMIPS:              3990.67
Virtualization:        VT-x
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              4096K
L3 cache:              16384K
NUMA node0 CPU(s):     0-47
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single pti ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat umip pku ospke md_clear arch_capabilities
```

2.  GCC 12
```
bash-4.2# gcc --version
gcc (GCC) 12.2.1 20221121 (Red Hat 12.2.1-4)
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

3. OS Info:
```
bash-4.2# cat /etc/os-release 
NAME="Oracle Linux Server"
VERSION="7.9"
ID="ol"
ID_LIKE="fedora"
VARIANT="Server"
VARIANT_ID="server"
VERSION_ID="7.9"
PRETTY_NAME="Oracle Linux Server 7.9"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:oracle:linux:7:9:server"
HOME_URL="https://linux.oracle.com/"
BUG_REPORT_URL="https://github.com/oracle/oracle-linux"

ORACLE_BUGZILLA_PRODUCT="Oracle Linux 7"
ORACLE_BUGZILLA_PRODUCT_VERSION=7.9
ORACLE_SUPPORT_PRODUCT="Oracle Linux"
ORACLE_SUPPORT_PRODUCT_VERSION=7.9
```
Let me know if any other details are needed. 

Looking forward to convert my PEFT model so that I can run efficiently on this CPU configuration.

Thanks for the support!

**Comment by zhentaoyu** [comment]
What is your model's `num_kv_heads`? The original [falcon-7b-instruct](https://huggingface.co/tiiuae/falcon-7b-instruct/blob/main/config.json) is **1**. And it should not run into #line132.
However, it seems your finetuned model's `num_kv_heads=71` which is the num of attention_heads.

https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py#L80

https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py#L131-L133 

What is your finetuned models' config.json looks like?
https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py#L70-L71

**Comment by amir1m** [comment]
Hi @zhentaoyu ,

1. Heads in base Falcon:

> _The original [falcon-7b-instruct](https://huggingface.co/tiiuae/falcon-7b-instruct/blob/main/config.json) is 1. And it should not run into #line132.
However, it seems your finetuned model's num_kv_heads=71 which is the num of attention_heads._

I see 71 heads even for the base falcon-7b-instruct. What am I missing?:
```
FalconConfig {
  "_name_or_path": "tiiuae/falcon-7b-instruct",
  "alibi": false,
  "apply_residual_connection_post_layernorm": false,
  "architectures": [
    "FalconForCausalLM"
  ],
  "attention_dropout": 0.0,
  "auto_map": {
    "AutoConfig": "tiiuae/falcon-7b-instruct--configuration_falcon.FalconConfig",
    "AutoModel": "tiiuae/falcon-7b-instruct--modeling_falcon.FalconModel",
    "AutoModelForCausalLM": "tiiuae/falcon-7b-instruct--modeling_falcon.FalconForCausalLM",
    "AutoModelForQuestionAnswering": "tiiuae/falcon-7b-instruct--modeling_falcon.FalconForQuestionAnswering",
    "AutoModelForSequenceClassification": "tiiuae/falcon-7b-instruct--modeling_falcon.FalconForSequenceClassification",
    "AutoModelForTokenClassification": "tiiuae/falcon-7b-instruct--modeling_falcon.FalconForTokenClassification"
  },
  "bias": false,
  "bos_token_id": 11,
  "eos_token_id": 11,
  "hidden_dropout": 0.0,
  "hidden_size": 4544,
  "initializer_range": 0.02,
  "layer_norm_epsilon": 1e-05,
  "model_type": "falcon",
  "multi_query": true,
  "new_decoder_architecture": false,
  "num_attention_heads": 71,
  "num_hidden_layers": 32,
  **"num_kv_heads": 71,**
  "parallel_attn": true,
  "quantization_config": {
    "bnb_4bit_compute_dtype": "bfloat16",
    "bnb_4bit_quant_type": "nf4",
    "bnb_4bit_use_double_quant": true,
    "llm_int8_enable_fp32_cpu_offload": false,
    "llm_int8_has_fp16_weight": false,
    "llm_int8_skip_modules": null,
    "llm_int8_threshold": 6.0,
    "load_in_4bit": true,
    "load_in_8bit": false
  },
  "torch_dtype": "bfloat16",
  "transformers_version": "4.31.0",
  "use_cache": true,
  "vocab_size": 65024
}
```

**2. PEFT Model details:**
```
PeftModelForCausalLM(
  (base_model): LoraModel(
    (model): FalconForCausalLM(
      (transformer): FalconModel(
        (word_embeddings): Embedding(65024, 4544)
        (h): ModuleList(
          (0-31): 32 x FalconDecoderLayer(
            (self_attention): FalconAttention(
              (maybe_rotary): FalconRotaryEmbedding()
              (query_key_value): Linear4bit(
                in_features=4544, out_features=4672, bias=False
                (lora_dropout): ModuleDict(
                  (default): Dropout(p=0.05, inplace=False)
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=4544, out_features=16, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=16, out_features=4672, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
              )
              (dense): Linear4bit(
                in_features=4544, out_features=4544, bias=False
                (lora_dropout): ModuleDict(
                  (default): Dropout(p=0.05, inplace=False)
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=4544, out_features=16, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=16, out_features=4544, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
              )
              (attention_dropout): Dropout(p=0.0, inplace=False)
            )
            (mlp): FalconMLP(
              (dense_h_to_4h): Linear4bit(
                in_features=4544, out_features=18176, bias=False
                (lora_dropout): ModuleDict(
                  (default): Dropout(p=0.05, inplace=False)
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=4544, out_features=16, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=16, out_features=18176, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
              )
              (act): GELU(approximate='none')
              (dense_4h_to_h): Linear4bit(
                in_features=18176, out_features=4544, bias=False
                (lora_dropout): ModuleDict(
                  (default): Dropout(p=0.05, inplace=False)
                )
                (lora_A): ModuleDict(
                  (default): Linear(in_features=18176, out_features=16, bias=False)
                )
                (lora_B): ModuleDict(
                  (default): Linear(in_features=16, out_features=4544, bias=False)
                )
                (lora_embedding_A): ParameterDict()
                (lora_embedding_B): ParameterDict()
              )
            )
            (input_layernorm): LayerNorm((4544,), eps=1e-05, elementwise_affine=True)
          )
        )
        (ln_f): LayerNorm((4544,), eps=1e-05, elementwise_affine=True)
      )
      (lm_head): Linear(in_features=4544, out_features=65024, bias=False)
    )
  )
)
```

**3. Config.json:**
```
{
  "auto_mapping": null,
  "base_model_name_or_path": "tiiuae/falcon-7b-instruct",
  "bias": "none",
  "fan_in_fan_out": false,
  "inference_mode": true,
  "init_lora_weights": true,
  "layers_pattern": null,
  "layers_to_transform": null,
  "lora_alpha": 32,
  "lora_dropout": 0.05,
  "modules_to_save": null,
  "peft_type": "LORA",
  "r": 16,
  "revision": null,
  "target_modules": [
    "dense_4h_to_h",
    "dense",
    "query_key_value",
    "dense_h_to_4h"
  ],
  "task_type": "CAUSAL_LM"
}
```

Thanks.

**Comment by zhentaoyu** [comment]
> I see 71 heads even for the base falcon-7b-instruct. What am I missing?:

1. The reason I think is from the code misalignment of `Falcon`.  `falcon-7b` uses multi-query attention and its `num_kv_heads=1`. see [here](https://huggingface.co/tiiuae/falcon-7b-instruct/blob/main/modeling_falcon.py#L222). But, its `config.json` didn't set this value. Then `AutoConfig` class init `num_kv_heads == num_attention_heads` , see [here](https://huggingface.co/tiiuae/falcon-7b-instruct/blob/main/configuration_falcon.py#L137).

2. I try to use [`config.json`](https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py#L70-L71) instead of  `AutoConfig`  for writing some important values for conversion to avoid this problem. However, I can not reproduce your results since I do have not your PEFT model. Maybe you can set the `n_head_kv=1` in your local `convert_falcon.py` and try again (this [line](https://github.com/intel/intel-extension-for-transformers/blob/main/intel_extension_for_transformers/llm/runtime/graph/scripts/convert_falcon.py#L80))?

**Comment by amir1m** [comment]
Hi @zhentaoyu ,
I followed your suggestion # 2 and changed the `n_head_kv=1`. After that I was able to run the convert and quantize successfully.  I am not sure what are the side effects. However after quantizing to INT8, the model hallucinated and was generating irrelevant text.

So I tried running `inference.py` on the converted (.bin f32) model instead of quantized one and it was generating text as expected. May be the irrelevant text was the side effect of the quantization, am not sure.

Thanks for the help!
